pkg/apis/pipeline/v1/pipeline_validation.go

```diff
@@ -497,6 +497,21 @@ func (pt *PipelineTask) extractAllParams() Params {
 	return allParams
 }
 
+func (pt *PipelineTask) GetVarSubstitutionExpressions() ([]string, bool) {
+	var allExpressions []string
+	if pt.TaskSpec != nil {
+		for _, step := range pt.TaskSpec.Steps {
+			stepExpressions, _ := step.GetVarSubstitutionExpressions()
+			allExpressions = append(allExpressions, stepExpressions...)
+		}
+		for _, sidecar := range pt.TaskSpec.Sidecars {
+			sidecarExpressions, _ := sidecar.GetVarSubstitutionExpressions()
+			allExpressions = append(allExpressions, sidecarExpressions...)
+		}
+	}
+	return allExpressions, len(allExpressions) != 0
+}
+
```

Comment:
I wonder if we should move the `whenExpression.GetVarSubstitutionExpressions()` and `GetVarSubstitutionExpressionsForParam(p)` as well

And also change `GetVarSubstitutionExpressionsForParam(p)` to `param.GetVarSubstitutionExpressions()`

But this is a separate cleanup

Reply:
Opened https://github.com/tektoncd/pipeline/pull/7121 to help with the cleanup, PTAL

Reply:
> I wonder if we should move the `whenExpression.GetVarSubstitutionExpressions()` and `GetVarSubstitutionExpressionsForParam(p)` as well
> 
> And also change `GetVarSubstitutionExpressionsForParam(p)` to `param.GetVarSubstitutionExpressions()`
> 
> But this is a separate cleanup

yeah, i agree with you. i add TODO in `PipelineTaskResultRefs`, i don’t want to add extra work to this PR, I  will submit the sparate cleanup PR later